internal/rbd/controllerserver.go

```diff
@@ -210,6 +210,39 @@ func validateRequestedVolumeSize(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnaps
 	return nil
 }
 
+func checkValidCreateVolumeRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials) error {
+	err := validateRequestedVolumeSize(rbdVol, parentVol, rbdSnap, cr)
+	if err != nil {
+		return err
+	}
+
+	return checkValidEncryptionRequest(parentVol, rbdVol, rbdSnap)
+}
+
+func checkValidEncryptionRequest(src, dest *rbdVolume, rbdSnap *rbdSnapshot) error {
```

Comment:
A snapshot is encrypted if its parent volume is encrypted. I do not think there is a need to pass `rbdSnap` here.

```go
func checkValidEncryptionRequest(src, dest *rbdImage) error {
```

and then call  this function only with the source (parentVol or rbdSnap) and the destination. Handling multiple cases in this function is more complex and confusing than is needed.

Reply:
we are passing dest and rbdSnap for two different CreateVolume cases. One is with PVC-PVC clone and Restore snapshot to a new PVC. not sure how I can manage with only src and dest here.

Reply:
Sure, but is there not always a single source and a single destination? In `checkValidCreateVolumeRequest` you can call `checkValidEncryptionRequest` depending on `parentVol` or `rbdSnap`? Use `rbdSnap` unless it is `nil`, otherwise use `parentVol`?

Reply:
In or other place i need to check for nil objects. I thought doing it in checkValidEncryptionRequest and send proper detailed error message? Do you have any suggestions how the function(pseudo code) should be like?

Reply:
maybe something like this:

```
func checkValidCreateVolumeRequest(rbdVol, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials) error {
        err := validateRequestedVolumeSize(rbdVol, parentVol, rbdSnap, cr)
        if err != nil {
                return err
        }

        if rbdSnap != nil {
                err = checkValidEncryptionRequest(&rbdSnap.rbdImage, &rbdVol.rbdImage)
        } else {
                err = checkValidEncryptionRequest(&parentVol.rbdImage, &rbdVol.rbdImage)
        }

        return err
}

// checkValidEncryptionRequest verifies that both src and dest have the same
// encryption status.
func checkValidEncryptionRequest(src, dest *rbdImage) error {
        if src != nil {
                switch {
                case src.isEncrypted() && !dest.isEncrypted():
                        return status.Errorf(codes.InvalidArgument, "cannot create unencrypted volume from encrypted volume")

                case !src.isEncrypted() && dest.isEncrypted():
                        return status.Errorf(codes.InvalidArgument, "cannot create encrypted volume from unencrypted volume")
                }
        }
        return nil
}
```

The error message should be clear in any case, the user would know if a clone or restore from snapshot is done.

Reply:
@nixpanic but with the above pseudo-code even in the case of Restore from snapshot it's returning a common error message which holds good for PVC-PVC clone.